# This patch file is from Qualcomm Technologies, Inc. and is provided under the following license:
#
# Copyright (c) 2025 Qualcomm Technologies, Inc. All rights reserved.
#
# SPDX-License-Identifier: BSD-3-Clause-Clear

From f5edd00d5fa1bd4f674bc2a7434c4af2d507bb86 Mon Sep 17 00:00:00 2001
From: Rafal Skierkowski <rskierko@qti.qualcomm.com>
Date: Thu, 21 Mar 2024 16:02:10 +0100
Subject: [PATCH 1/2] Handle outer device tree case in boot image.

When "outer" device tree type is enabled in coqoshv guest configuration,
imageForge is placing device tree outside of HV image and set
DeviceTreeLoadAddr as per guest configuration. In our case, device tree
placement is lower then KernelLoadAddr (here it is not kernel but hypervisor).
This is detected when DeviceTreeLoadAddr is on lower address then
KernelLoadAddr. Rework bootloader logic in that case to avoid overflow
error check, but still keep compatibility with previous approach.

This functionality is to allow bootloader to pass modified device tree
to Linux VM.

Signed-off-by: Ankurkumar Satasiya <asatasiy@qti.qualcomm.com>
---
 .../QcomModulePkg/Library/BootLib/BootLinux.c | 23 ++++++++++++-------
 1 file changed, 15 insertions(+), 8 deletions(-)

diff --git a/bootloader/edk2/QcomModulePkg/Library/BootLib/BootLinux.c b/bootloader/edk2/QcomModulePkg/Library/BootLib/BootLinux.c
index ce20b24..65c2fc8 100644
--- a/QcomModulePkg/Library/BootLib/BootLinux.c
+++ b/QcomModulePkg/Library/BootLib/BootLinux.c
@@ -634,7 +634,7 @@ STATIC EFI_STATUS
 GZipPkgCheck (BootParamlist *BootParamlistPtr)
 {
   UINT32 OutLen = 0;
-  UINT64 OutAvaiLen = 0;
+  UINT64 OutAvaiLen = MAX_UINT32;
   struct kernel64_hdr *Kptr = NULL;
 
   if (BootParamlistPtr == NULL) {
@@ -644,14 +644,21 @@ GZipPkgCheck (BootParamlist *BootParamlistPtr)
   }
 
   if (BootParamlistPtr->BootingWithGzipPkgKernel) {
-    OutAvaiLen = BootParamlistPtr->DeviceTreeLoadAddr -
-                 BootParamlistPtr->KernelLoadAddr;
+    if (BootParamlistPtr->DeviceTreeLoadAddr > BootParamlistPtr->KernelLoadAddr) {
+      OutAvaiLen = BootParamlistPtr->DeviceTreeLoadAddr -
+                   BootParamlistPtr->KernelLoadAddr;
 
-    if (OutAvaiLen > MAX_UINT32) {
-      DEBUG ((EFI_D_ERROR,
-              "Integer Overflow: the length of decompressed data = %u\n",
-      OutAvaiLen));
-      return EFI_BAD_BUFFER_SIZE;
+      if (OutAvaiLen > MAX_UINT32) {
+        DEBUG ((EFI_D_ERROR,
+                "Integer Overflow: the length of decompressed data = %u\n",
+        OutAvaiLen));
+        return EFI_BAD_BUFFER_SIZE;
+      }
+    } else {
+      DEBUG ((EFI_D_INFO,
+              "DeviceTreeLoadAddr %x is lower then KernelLoadAddr %x\n",
+              BootParamlistPtr->DeviceTreeLoadAddr,
+              BootParamlistPtr->KernelLoadAddr));
     }
 
     DEBUG ((EFI_D_INFO, "Decompressing kernel image start: %lu ms\n",
-- 
2.43.2

